In [1]:
#Read data directly from Athena
from pyathena import connect
import pandas as pd
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules
import mlxtend as ml

In [2]:
###!pip install mlxtend

In [3]:
ACCESS_KEY = "AKIAU7APHSU5V6VGK7XV"
AWS_SECRET_ACCESS_KEY = "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"
STAGING_S3_DIR = "s3://aws-athena-query-results-341481854267-eu-west-1/"

In [4]:
conn = connect(aws_access_key_id=ACCESS_KEY,
               aws_secret_access_key=AWS_SECRET_ACCESS_KEY,
               s3_staging_dir=STAGING_S3_DIR,
               region_name='eu-west-1')


In [5]:
dfFormParquet = pd.read_sql("SELECT * FROM app_db_p.formsparquet", conn)
dfSignUpTable = pd.read_sql("SELECT * FROM app_db_p.sign_up_parquet", conn)

In [6]:
print(dfSignUpTable.shape)
print(dfFormParquet.shape)

(1933, 49)
(3171, 16)


In [ ]:
dfSignUpTable.head()

In [ ]:
dfFormParquet.head()

In [21]:
data = dfFormParquet.groupby(["form_type", "purpose"]).size().reset_index(name='count')

In [30]:
transformed = data.groupby(["form_type", "purpose"])['count'].sum().unstack().reset_index().fillna(0)

In [29]:
def encode_units(x):
    if x <= 0:
        return 0
    if x >= 1:
        return 1

In [33]:
transformed.head()

purpose   form_type  Application  Banking  Census  Contact  Education  \
0        BIOMETRICS           10        7       8        6          7   
1           WEBFORM           26       11       5        6          7   

purpose  Feedback  File upload forms  Human resource  Order  Others  \
0               7                  8               3      1       5   
1              15                  7               6      5      35   

purpose  Questionnaire  Registration  Request  Research  
0                    2            18        4         8  
1                   15            84       13         4

In [36]:
transformed['form_type'] = transformed['form_type'].astype('category')

transformed['form_type'] = transformed['form_type'].cat.codes

datapresent = transformed.applymap(encode_units)

frequent_itemsets = apriori(datapresent, min_support=0.01, use_colnames=True)

rules = association_rules(frequent_itemsets, metric='lift')

#### Differential Customer Analysis.

In [30]:
#How to make differential customer analysis
#How to assign if a customer subscribed or not.

subscribedClients = dfSignUpTable[(dfSignUpTable.subscription_date - dfSignUpTable.sign_up_date) > pd.Timedelta(' 14 days')]

In [32]:
subscribedClients.columns

Index(['createdby', 'email', 'customer_name', 'company', 'industry',
       'sign_up_date', 'orgid', 'customer_source', 'sign_up_template',
       'num_forms', 'number_of_users', 'number_of_published_forms', 'phone',
       'country', 'role', 'capturd_records', 'template_name', 'purpose',
       'form_type', 'subscription_date', 'p', 'userid', 'entry_date',
       'contact_date', 'customer_succes_exec', 'customer_name_',
       'company_name', 'email_address', 'phone_number', 'country_',
       'how_did_u_hear_abt_us', 'refered_by', 'reason_for_signup',
       'prefered_form_type', 'intended_use_of_data', 'customer_feedback',
       'call_comments', 'further_engagement', 'lead_pused_to_sales',
       'which_person', 'next_followup_date', 'followup_feedback',
       'followup_feedback_date', 'followup_feedback_2',
       'followup_feedback_date_2', 'customer_succes_exec_feedback2',
       'followup_feedback_3', 'followup_feedback_date_3',
       'customer_succes_exec_feedback3'],
      

In [ ]:
subscribedClients.how_did_u_hear_abt_us

In [ ]:
subscribedClients.customer_source

In [ ]:
subscribedClients.industry

In [ ]:
subscribedClients.country

In [ ]:
subscribedClients.template_name

In [37]:
#How many days have this guys spent from sign up
subscribedClients['days_spent'] = subscribedClients.subscription_date - subscribedClients.sign_up_date
subscribedClients.days_spent

/Users/adekunleba/anaconda3/envs/tensorflow13/lib/python3.7/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  


4       15 days 09:15:17
75     118 days 18:18:45
231     20 days 21:06:15
318     57 days 22:44:35
382     28 days 02:40:37
435     28 days 01:56:47
471     91 days 16:36:49
472     91 days 16:36:49
473     91 days 16:36:49
843     95 days 02:00:04
894    177 days 13:10:35
979     33 days 09:48:55
1179   240 days 02:19:15
1248   161 days 18:44:36
1323    26 days 01:42:11
1431   323 days 04:01:57
1463   186 days 12:36:37
1818    14 days 17:34:57
1866    27 days 11:51:46
Name: days_spent, dtype: timedelta64[ns]

In [40]:
# Delving deep into financial industry

financialInstitutionData = dfSignUpTable[dfSignUpTable.industry == "FINANCIAL INSTITUTION"]

In [ ]:
financialInstitutionData['days_spent'] = financialInstitutionData.subscription_date - financialInstitutionData.sign_up_date
financialInstitutionData.days_spent

In [46]:
#What percent stayed above 80 days i.e 75% maximum stayed date
subscribedClients[subscribedClients.days_spent > pd.Timedelta("80 days")].shape[0] / subscribedClients.shape[0]

0.5263157894736842

In [ ]:
fcmgIndustry = dfSignUpTable[dfSignUpTable.industry == "FMCG"]
fcmgIndustry['days_spent'] = fcmgIndustry.subscription_date - fcmgIndustry.sign_up_date
fcmgIndustry.days_spent

In [ ]:
#Analysis on how did you hear about us.
faceBookSignUps = dfSignUpTable[dfSignUpTable.how_did_u_hear_abt_us == 'Facebook']
faceBookSignUps['days_spent'] = faceBookSignUps.subscription_date - faceBookSignUps.sign_up_date
faceBookSignUps[faceBookSignUps.days_spent != pd.Timedelta('0 days')]['days_spent']

In [ ]:
seminarSignUps = dfSignUpTable[dfSignUpTable.how_did_u_hear_abt_us == 'Seminar']
seminarSignUps['days_spent'] = seminarSignUps.subscription_date - seminarSignUps.sign_up_date
seminarSignUps[seminarSignUps.days_spent != pd.Timedelta('0 days')]['days_spent']

##### Some questions asked
- The question is? is it worth it to pursue a specific industry
- What are the characteristics of customers in these worthy industries.

- Another differential analysis will be should we pursue companies with a specific source of acquisition.
- What are the characteristics of the worthy customer source.

- Another differential analysis will be are the templates choosen for sign up, any contributor.

##### Comments

* The only yielding sources of acquisition thus far is `Facebook` and `Seminar`. However, there isn't much confidence though who heard from `Facebook` compared to those who heard from `Seminar`. Because overall **0.6%** of those that heard from Facebook actually became a paying customer compared to those that heard from `Seminar` which was a full 100% although, we still need to solidfy this claim further as there was a little class imbalance that was overlooked for this claim.
* It is also further important to keep pushing for the `Financial Institutions` and the `Fast moving consumer goods` as they do show potential as probable industry that will likely yield high converstion, however this is subject to further analysis on substantial data. This caution is so because, the client from the financial industry and fmcg only spent 28 days and 14 days respectively while the longest staying person has stayed up to 323 days and  52% of the client has stayed past 75% of the longest staying time.

In [60]:
(1/164) * 100

0.6097560975609756

In [64]:
m = [0] * 1000
m[8]

In [65]:
m[8]

1

In [68]:
!ls -lh

total 80
-rw-r--r--  1 adekunleba  staff    28K Oct  2 13:44 MarketBasketAnalysis.ipynb
-rw-r--r--  1 adekunleba  staff   7.1K Sep 24 11:09 Untitled.ipynb
